Plugins/Public/dronebays/UserCommands.cpp

```diff
@@ -0,0 +1,377 @@
+		const uint targetid = HkGetClientIDByShip(clientDroneInfo[iClientID].deployedInfo.lastShipObjTarget);
+		const wstring targetname = reinterpret_cast<const wchar_t*>(Players.GetActiveCharacterName(targetid));
+
+		// Only bother logging if we weren't engaging a NPC
+		if (!targetname.empty())
+		{
+			wstring logString = L"Player %s has ordered its drone to target %t";
+			logString = ReplaceStr(logString, L"%s", charname);
+			logString = ReplaceStr(logString, L"%t", targetname);
+			Utility::LogEvent(wstos(logString).c_str());
+
+			// Since we know this was a real player, alert them that they're being engaged
+			PrintUserCmdText(targetid, L"Player %s has targeted you with a drone!", charname);
+		}
+	}
+
+	return true;
+}
+
+
+bool UserCommands::UserCmd_RecallDrone(uint iClientID, const wstring& wscCmd, const wstring& wscParam,
+	const wchar_t* usage)
+{
+	// Verify that the user is in space
+	uint iShipObj;
+	pub::Player::GetShip(iClientID, iShipObj);
+	if (!iShipObj)
+	{
+		PrintUserCmdText(iClientID, L"You must be in space to use this command");
+		return true;
+	}
+
+	// Verify that the user has a drone currently deployed
+	if (clientDroneInfo[iClientID].deployedInfo.deployedDroneObj == 0)
+	{
+		PrintUserCmdText(iClientID, L"You must have a drone deployed for this to work");
+		return true;
+	}
+
+	// Set the NPC to fly to your current position
+	pub::AI::DirectiveGotoOp gotoOp;
+	
+	// Type zero says to fly to a spaceObj
+	gotoOp.iGotoType = 0;
+	gotoOp.iTargetID = iShipObj;
+	gotoOp.fRange = 10.0;
+	gotoOp.fThrust = 100;
+	gotoOp.goto_cruise = true;
+
+	pub::AI::SubmitDirective(clientDroneInfo[iClientID].deployedInfo.deployedDroneObj, &gotoOp);
+
+	//Create the timer entry to keep watch on this docking operation
+	DroneDespawnWrapper wrapper;
+	wrapper.droneObj = clientDroneInfo[iClientID].deployedInfo.deployedDroneObj;
+	wrapper.parentObj = iShipObj;
+	
+	droneDespawnMap[iClientID] = wrapper;
+
+	PrintUserCmdText(iClientID, L"Drone recall operation sent");
+
+	return true;
+}
+
+bool UserCommands::UserCmd_DroneStop(uint iClientID, const wstring& wscCmd, const wstring& wscParam,
+	const wchar_t* usage)
+{
+	// Verify that the user is in space
+	uint iShipObj;
+	pub::Player::GetShip(iClientID, iShipObj);
+	if (!iShipObj)
+	{
+		PrintUserCmdText(iClientID, L"You must be in space to use this command");
+		return true;
+	}
+
+	// Verify that the user has a drone currently deployed
+	if (clientDroneInfo[iClientID].deployedInfo.deployedDroneObj == 0)
+	{
+		PrintUserCmdText(iClientID, L"You must have a drone deployed for this to work");
+		return true;
+	}
+
+	const uint droneObj = clientDroneInfo[iClientID].deployedInfo.deployedDroneObj;
+
+	// Set the drone reputation to neutral with who it was last attacking
+	Utility::SetRepNeutral(droneObj, clientDroneInfo[iClientID].deployedInfo.lastShipObjTarget);
+
+	// Send a stop directive to the drone
+	pub::AI::DirectiveCancelOp cancelOp;
+	SubmitDirective(droneObj, &cancelOp);
```

Comment:
pub::AI:: ?

Reply:
Explicit namespace is in there. Not required. Doesn't hurt. Leave it as a reminder where the function is.

Reply:
explicit namespace is not in there, it is missing

Reply:
.

Reply:
fixed in my next commit